src/components/heatmap/heatmap.tsx

```diff
@@ -0,0 +1,344 @@
+      if (!configErrorsPresent) {
+        const sortBy = typeof config.data.sortBy !== 'undefined' ? config.data.sortBy : [];
+        const ref = reference.sort(sortBy);
+        setHeader((prevHeader) => ({ ...prevHeader, action: 'main' }));
+        const pcid = getQueryParams(location.href).pcid
+        const ppid = getQueryParams(location.href).ppid
+        if (pcid)
+          setHeader((prevHeader) => ({
+            ...prevHeader,
+            pcid,
+          }));
+        if (ppid)
+          setHeader((prevHeader) => ({
+            ...prevHeader,
+            ppid,
+          }))
+        ref.read(1000, header).then(function getPage(page: any) {
+          readAll(page);
+        }).catch(function () {
+          setInvalidConfigs(['Error while reading data from Ermrestjs']);
+          setConfigErrorsPresent(true);
+        });
+      }
+    });
+
+  }, []);
+
+  useEffect(() => {
+    setFacet({
+      'and': [{
+        'markdown_name': 'Gene NCBI ID',
+        'entity': false,
+        'choices': [NCBIGeneID],
+        'source': 'NCBI_GeneID',
+        'hide_not_null_choice': true,
+        'hide_null_choice': true
+      }]
+    });
+  }, [NCBIGeneID]);
+
+
+  const verifyConfiguration = (reference: any) => {
+    const columns = ['titleColumn', 'idColumn', 'xColumn', 'yColumn', 'zColumn'];
+    setInvalidConfigs(() => []);
+    for (let i = 0; i < columns.length; i++) {
+      try {
+        reference.getColumnByName(config.data[columns[i]]);
+      } catch (error) {
+        setInvalidConfigs((prevInvalidConfigs) =>
+          ([...prevInvalidConfigs, 'Coulmn \'' + config.data[columns[i]] + '\' does not exist. Give a valid value for the ' + columns[i] + '.']))
+      }
+    }
+    const sortColumns = config.data.sortBy;
+    for (let i = 0; i < sortColumns.length; i++) {
+      try {
+        reference.getColumnByName(sortColumns[i].column);
+      } catch (error) {
+        setInvalidConfigs((prevInvalidConfigs) =>
+          ([...prevInvalidConfigs, 'Coulmn \'' + sortColumns[i].column + '\' in \'sortBy\' field does not exist. Replace it with a valid column.']));
+      }
+    }
+    if (invalidConfigs.length > 0) {
+      setInvalidConfigs(() => invalidConfigs);
+      setConfigErrorsPresent(true);
+    }
+  }
+
+  /**
+   * @param page : Page object fecthed from response of ermrestURI
+   */
+  const readAll = (page: any) => {
+    for (let i = 0; i < page.tuples.length; i++) {
+      addData(page.tuples[i]);
+    }
+    if (page.hasNext) {
+      Object(header).action = 'page';
+      page.next.read(1000, header).then(readAll).catch(function () {
+        setInvalidConfigs(() => ['Error while reading data from Ermrestjs']);
+        setConfigErrorsPresent(() => false);
+      });
+    } else {
+      setHeatmaps(() => (heatmaps));
+    }
+    setNCBIGeneID(() => (page.tuples[0].data.NCBI_GeneID));
+  }
+  /**
+   * 
+   * @param tuple : Tuple object retrieved from page
+   * Adds data(x,y,z values) to heatmap objects
```

Comment:
Can you add more details about how this is adding the data?

For instance, is this adding 1 y and 1 x value if they aren't in the set yet? And then the corresponding z values?